src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,59 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
+            int recommendedSize = int.MaxValue / 4 > source.Length ? source.Length * 4 : source.Length;
+
+            // "Call ucol_getSortKey() to find out how big the sort key buffer should be, and fill in the temporary buffer at the same time"
+            if (TryGetSortKey(source, options, recommendedSize, out int actualSortKeyLength, out int hash))
             {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+                return hash; // hot path, the buffer was big enough and we don't need to call the method one more time
+            }
 
-                byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
+            // "If the temporary buffer is too small, allocate or reallocate more space for in anoverflow buffer to handle the overflow situations."
+            // "If there was an internal error generating the sort key, a zero value is returned. "
+            if (actualSortKeyLength > 0 && TryGetSortKey(source, options, actualSortKeyLength, out _, out hash))
+            {
+                return hash;
+            }
 
-                fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
-                {
-                    if (Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, sortKeyLength, options) != sortKeyLength)
-                    {
-                        throw new ArgumentException(SR.Arg_ExternalException);
-                    }
-                }
+            throw new ArgumentException(SR.Arg_ExternalException);
+        }
+        
+        private unsafe bool TryGetSortKey(ReadOnlySpan<char> source, CompareOptions options, int requestedSortKeyLength, out int actualSortKeyLength, out int hash)
+        {
+            Debug.Assert(requestedSortKeyLength > 0);
+            Debug.Assert(source.Length > 0);
 
-                int hash = Marvin.ComputeHash32(span.Slice(0, sortKeyLength), Marvin.DefaultSeed);
+            byte[]? borrowedArr = null;
+            Span<byte> span = requestedSortKeyLength <= 512 ?
+                stackalloc byte[requestedSortKeyLength] :
+                (borrowedArr = ArrayPool<byte>.Shared.Rent(requestedSortKeyLength));
 
-                // Return the borrowed array if necessary.
-                if (borrowedArr != null)
-                {
-                    ArrayPool<byte>.Shared.Return(borrowedArr);
-                }
+            fixed (char* pSource = &MemoryMarshal.GetReference(source))
+            fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
+            {
+                actualSortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, requestedSortKeyLength, options);
```

Comment:
Use the actual size of the Span here instead of requestedSortKeyLength. The ArrayPool often gives you larger array than what you asked for. It would be a shame to not use the extra space.

Reply:
good idea